Reply/app/src/main/java/com/example/reply/ui/navigation/ReplyNavigationComponents.kt

```diff
@@ -49,20 +50,119 @@ import androidx.compose.material3.NavigationRail
+import androidx.compose.material3.adaptive.navigationsuite.NavigationSuiteScaffoldLayout
+import androidx.compose.material3.adaptive.navigationsuite.NavigationSuiteType
+import androidx.compose.material3.rememberDrawerState
 import androidx.compose.runtime.Composable
+import androidx.compose.runtime.rememberCoroutineScope
 import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
 import androidx.compose.ui.graphics.Color
 import androidx.compose.ui.layout.Layout
 import androidx.compose.ui.layout.Measurable
 import androidx.compose.ui.layout.MeasurePolicy
 import androidx.compose.ui.layout.layoutId
+import androidx.compose.ui.platform.LocalDensity
 import androidx.compose.ui.res.stringResource
 import androidx.compose.ui.text.style.TextAlign
 import androidx.compose.ui.unit.dp
 import androidx.compose.ui.unit.offset
+import androidx.compose.ui.unit.toSize
+import androidx.window.core.layout.WindowHeightSizeClass
+import androidx.window.core.layout.WindowSizeClass
+import androidx.window.core.layout.WindowWidthSizeClass
 import com.example.reply.R
 import com.example.reply.ui.utils.ReplyNavigationContentPosition
+import kotlinx.coroutines.launch
+
+private fun WindowSizeClass.isCompact() =
+    windowWidthSizeClass == WindowWidthSizeClass.COMPACT ||
+        windowHeightSizeClass == WindowHeightSizeClass.COMPACT
+
+class ReplyNavSuiteScope(
+    val navSuiteType: NavigationSuiteType
+)
+
+@Composable
+fun ReplyNavigationWrapper(
+    selectedDestination: String,
+    navigateToTopLevelDestination: (ReplyTopLevelDestination) -> Unit,
+    content: @Composable ReplyNavSuiteScope.() -> Unit
+) {
+    val adaptiveInfo = currentWindowAdaptiveInfo()
+    val windowSize = with(LocalDensity.current) {
+        currentWindowSize().toSize().toDpSize()
+    }
+
+    val navLayoutType = when {
+        adaptiveInfo.windowPosture.isTabletop -> NavigationSuiteType.NavigationBar
+        adaptiveInfo.windowSizeClass.isCompact() -> NavigationSuiteType.NavigationBar
+        adaptiveInfo.windowSizeClass.windowWidthSizeClass == WindowWidthSizeClass.EXPANDED &&
+            windowSize.width >= 1200.dp -> NavigationSuiteType.NavigationDrawer
+        else -> NavigationSuiteType.NavigationRail
+    }
+    val navContentPosition = when (adaptiveInfo.windowSizeClass.windowHeightSizeClass) {
+        WindowHeightSizeClass.COMPACT -> ReplyNavigationContentPosition.TOP
+        WindowHeightSizeClass.MEDIUM,
+        WindowHeightSizeClass.EXPANDED -> ReplyNavigationContentPosition.CENTER
+        else -> ReplyNavigationContentPosition.TOP
+    }
+
+    val drawerState = rememberDrawerState(initialValue = DrawerValue.Closed)
+    val coroutineScope = rememberCoroutineScope()
+    // Avoid opening the modal drawer when there is a permanent drawer or a bottom nav bar,
+    // but always allow closing an open drawer.
+    val gesturesEnabled =
+        drawerState.isOpen || navLayoutType == NavigationSuiteType.NavigationRail
```

Comment:
This logic does makes sense for consistency, where the drawer will remain open if it was open, but maybe an alternative would be to just hide it immediately if the navigation layout type isn't a navigation rail? Not sure which way would be better.

Reply:
I tried something like that, and it looked fine when changing to a layout that isn't a rail. But when changing back to the rail the drawer opens again, and that's a bit jarring.

We can work around that with a `LaunchedEffect` that calls `drawerState.snapTo(Closed)`, but that feels more imperative than declarative.